Source/GSCSConstraint.m

```diff
@@ -1,27 +1,366 @@
+}
+
+- (instancetype) initEditConstraintWithVariable: (GSCSVariable *)variable
+                                        stength: (GSCSStrength *)strength
+{
+  GSCSLinearExpression *expression =
+      [[GSCSLinearExpression alloc] initWithVariable: variable
+                                         coefficient: -1
+                                            constant: [variable value]];
+  GSCSConstraint *constraint = [self initWithType: GSCSConstraintTypeEdit
+                   strength: strength
+                 expression: expression
+                   variable: variable];
+  RELEASE(expression);
+  return constraint;
+}
+
+- (instancetype) initStayConstraintWithVariable: (GSCSVariable *)variable
+                                       strength: (GSCSStrength *)strength
+{
+  GSCSLinearExpression *expression =
+      [[GSCSLinearExpression alloc] initWithVariable: variable
+                                         coefficient: -1
+                                            constant: [variable value]];
+  GSCSConstraint *constraint = [[GSCSConstraint alloc] initWithType: GSCSConstraintTypeStay
+                                     strength: strength
+                                   expression: expression
+                                     variable: variable];
+  RELEASE(expression);
+  return constraint;
+}
+
+- (instancetype) initWithLhsVariable: (GSCSVariable *)lhs
+                   equalsRhsVariable: (GSCSVariable *)rhs
+{
+  GSCSLinearExpression *expression =
+      [[GSCSLinearExpression alloc] initWithVariable: lhs];
+  [expression addVariable: rhs coefficient: -1];
+  GSCSConstraint *constraint = [self initLinearConstraintWithExpression: expression];
+  RELEASE(expression);
+
+  return constraint;
+}
+
+- (instancetype) initWithLhsVariable: (GSCSVariable *)lhs
+                      equalsConstant: (CGFloat)rhs
+{
+  GSCSLinearExpression *expression =
+      [[GSCSLinearExpression alloc] initWithVariable: lhs
+                                         coefficient: 1
+                                            constant: -rhs];
+  GSCSConstraint *constraint = [self initLinearConstraintWithExpression: expression];
+  RELEASE(expression);
+  return constraint;
+}
+
++ (GSCSConstraint *) constraintWithLeftVariable: (GSCSVariable *)lhs
+operator: (GSCSConstraintOperator) operator
+                                  rightVariable: (GSCSVariable *)rhsVariable
+{
+  if (operator== GSCSConstraintOperatorEqual)
+    {
+      GSCSLinearExpression *expression =
+          [[GSCSLinearExpression alloc] initWithVariable: lhs];
+      [expression addVariable: rhsVariable coefficient: -1];
+
+      RELEASE(expression);
+
+      return AUTORELEASE([[GSCSConstraint alloc]
+          initLinearConstraintWithExpression: expression]);
+    }
+  else
+    {
+      GSCSLinearExpression *rhsExpression =
+          [[GSCSLinearExpression alloc] initWithVariable: rhsVariable];
+      if (operator== GSCSConstraintOperationGreaterThanOrEqual)
```

Comment:
To me this duplication of code looks like we are missing out on having helper methods on GSCSLinearExpression to create different expressions for operators.

Reply:
I looked at this, a lot of these convenience methods are symmetrical but have minor differences. I'm not seeing refactoring opportunities that are worth the squeeze. Can you provide a suggestion of where you see one.

Reply:
You could always call the symmetrical method and invert the operator. 

GSCSConstraintOperationGreaterThanOrEqual => GSCSConstraintOperatorLessThanOrEqual
GSCSConstraintOperatorLessThanOrEqual => GSCSConstraintOperationGreaterThanOrEqual
GSCSConstraintOperatorEqual => GSCSConstraintOperatorEqual